From 26ae7e1ff628f203b8937b40b6d0e2584a12e6dd Mon Sep 17 00:00:00 2001
From: Jason Wang <jasowang@redhat.com>
Date: Mon, 29 Sep 2014 00:59:07 -0400
Subject: [virt] hyperv: Properly handle checksum offload

Message-id: <1411952358-34104-24-git-send-email-jasowang@redhat.com>
Patchwork-id: 96776
O-Subject: [RHEL7.1 PATCH 23/34] hyperv: Properly handle checksum offload
Bugzilla: 1146357
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Vitaly Kuznetsov <vkuznets@redhat.com>

From: KY Srinivasan <kys@microsoft.com>

Bugzilla: 1146357

Do checksum offload only if the client of the driver wants checksum to be
offloaded.

In V1 version of this patch, I  addressed comments from
Stephen Hemminger <stephen@networkplumber.org> and
Eric Dumazet <eric.dumazet@gmail.com>.

In this version of the patch I have addressed comments from
David Miller.

This patch fixes a bug that is exposed in gateway scenarios.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Reviewed-by: Haiyang Zhang <haiyangz@microsoft.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 22041fb05b66387b2854f789d1e1f55c7d07b4f4)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/hyperv/netvsc_drv.c b/drivers/net/hyperv/netvsc_drv.c
index f059d87..6b737ef 100644
--- a/drivers/net/hyperv/netvsc_drv.c
+++ b/drivers/net/hyperv/netvsc_drv.c
@@ -467,6 +467,10 @@ static int netvsc_start_xmit(struct sk_buff *skb, struct net_device *net)
  if (skb_is_gso(skb))
   goto do_lso;
 
+ if ((skb->ip_summed == CHECKSUM_NONE) ||
+     (skb->ip_summed == CHECKSUM_UNNECESSARY))
+  goto do_send;
+
  rndis_msg_size += NDIS_CSUM_PPI_SIZE;
  ppi = init_ppi_data(rndis_msg, NDIS_CSUM_PPI_SIZE,
        TCPIP_CHKSUM_PKTINFO);
-- 
1.7.1